### PR TITLE
test: Separate test and build workflows

### LINKED ISSUE #11: GitHub action failing with "No space left on device" error
https://github.com/IBM/text-generation-inference/actions/runs/6501966098/job/17660182367#step:4:2283

```
#30 175.9 ERROR: Could not install packages due to an OSError: [Errno 28] No space left on device
```

_Originally posted by @ckadner in https://github.com/IBM/text-generation-inference/issues/10#issuecomment-1760532996_

### PR DESCRIPTION
**Pull request by ckadner**
Subset of PR #12 -- focusing just on the CI parts

**Changes in this PR:**

- [x] add GitHub action to free up disk space (see results below)
- [x] split build and test into separate workflows

**Test:**
- [x] run Python tests and integration tests in parallel
- [x] build test image once, reuse it for Python tests and integration tests (`check-test-image` vs `build-test-image`)
- [x] ~~temporarily exclude/skip failing Python tests~~

**Build:**
- [x] temporarily build all stages of Dockerfile sequentially in order to be able to capture build logs
- [ ] separate flash-attention v1 build from rotary-embeddings and dropout-layer-norm builds
- [x] limit number of parallel jobs for compilation of `flash-attention` v2 to avoid OOM error
- [x] ~~upgrade flash-attention v2 to 2.5.2 to get pre-built wheels for PyTorch 2.3~~
- [x] ~~use the same PyTorch 2.3 nightly build as used by pre-built wheels for flash-attn-v2~~
- [x] build on push to main
- [x] add build status badge to README.md

**Results:**

- Disk usage before cleanup:
  ```
  Filesystem      Size  Used Avail Use% Mounted on
  /dev/root        84G   66G   18G  80% /
  ```
- Disk usage after:
  ```
  Filesystem      Size  Used Avail Use% Mounted on
  /dev/root        84G   33G   51G  40% /
  ```

Resolves #11

---

**Side note:**
- building the wheel for the `flash-attention` extension module dropout `layer_norm` takes about 1 hour (`3702` sec)
  ```
  #17 3702.2   Created wheel for dropout_layer_norm: filename=dropout_layer_norm-0.1-cp311-cp311-linux_x86_64.whl   size=149988151
  ```
- but it has been replaced as of 2024-01-05 in [v2.4.3](https://github.com/Dao-AILab/flash-attention/commit/abbc131)
- the FlashAttention repo switched to a [Triton-based implementation](https://github.com/Dao-AILab/flash-attention/blob/main/flash_attn/ops/triton/layer_norm.py)
- so the TGIS build could potentially be speed up by an hour

### CODE REVIEW

**Comment by ckadner** [comment]
@tjohnson31415

**Review by tjohnson31415 — COMMENTED** [review]
A couple tiny things

File: Dockerfile [hunk 1]
```diff
@@ -200,17 +205,19 @@ ENV PATH=/opt/tgis/bin/:$PATH
...
 # Install specific version of torch
 RUN pip install ninja==1.11.1.1 --no-cache-dir
+RUN pip install packaging --no-cache-dir
 RUN pip install torch==$PYTORCH_VERSION+cu118 --index-url "${PYTORCH_INDEX}/cu118" --no-cache-dir
 
 
 ## Build flash attention v2 ####################################################
 FROM python-builder as flash-att-v2-builder
-ARG FLASH_ATT_VERSION=v2.3.6
+ARG FLASH_ATT_VERSION=v2.5.2
```
[the remark this answers is not in the document]
Comment (ckadner) [reply Dockerfile]: interesting, I had reverted to `ARG FLASH_ATT_VERSION=v2.3.6` (i.e. no change) -- did you comment on an old commit?

File: .github/workflows/test.yml [hunk 2]
```diff
@@ -1,37 +1,120 @@
...
+            -e HUGGINGFACE_HUB_CACHE=/transformers_cache \
+            -e TRANSFORMERS_CACHE=/transformers_cache \
+            $TEST_IMAGE_NAME pytest -sv \
+              --ignore=server/tests/test_utils.py \
+              --ignore=server/tests/models/test_bloom.py \
+              --ignore=server/tests/models/test_causal_lm.py \
+              --ignore=server/tests/models/test_seq2seq_lm.py \
+              --ignore=server/tests/utils/test_convert.py \
+              -k "not cache_with_eviction and not tensor_too_large" \
+              server/tests
```
[the remark this answers is not in the document]
Comment (ckadner) [reply .github/workflows/test.yml:92]: I will give it a try

File: .github/workflows/build.yml [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+env:
+  CI: true
+  DOCKER_BUILDKIT: 1
+  SERVER_IMAGE_NAME: "text-gen-server:0"
+
+jobs:
+  build:
+    runs-on: ubuntu-latest
+    env:
+      # BUILDKIT_PROGRESS: plain
```
[the remark this answers is not in the document]
Comment (ckadner) [reply .github/workflows/build.yml:31]: agreed

File: .github/workflows/test.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (ckadner) [reply .github/workflows/test.yml]: works

**Comment by ckadner** [comment]
@njhill I merged in the latest changes from main and made a few more changes as per your review suggestions @tjohnson31415 

Could you give it one last look so we can merge?

**Review by tjohnson31415 — APPROVED** [review]
LGTM now!

**Pull request merged**